[oneDPL] Add C++ standard like execution policies info (#450)
oneapi-src/oneAPI-spec: source/elements/oneDPL/source/parallel_api.rst

execution policies*.

oneDPL extends the C++ Standard like API with `SYCL`_ buffer wrappers, iterators and additional
parallel algortithms.

C++ Standard aligned execution policies
+++++++++++++++++++++++++++++++++++++++

oneDPL has its own set of execution policies and utilities that represent functionality aligned with
C++ Standard:

.. code:: cpp

  // Defined in <oneapi/dpl/execution>

  namespace oneapi {
    namespace dpl {
      namespace execution {

        class sequenced_policy { /*unspecified*/ };
        class parallel_policy { /*unspecified*/ };
        class parallel_unsequenced_policy { /*unspecified*/ };
        class unsequenced_policy { /*unspecified*/ };

        inline constexpr sequenced_policy seq { /*unspecified*/ };
        inline constexpr parallel_policy par { /*unspecified*/ };
        inline constexpr parallel_unsequenced_policy par_unseq { /*unspecified*/ };
        inline constexpr unsequenced_policy unseq { /*unspecified*/ };

        template <class T>
        struct is_execution_policy;

        template <class T>
        inline constexpr bool is_execution_policy_v = oneapi::dpl::execution::is_execution_policy<T>::value;
      }
    }
  }

> Please see C++ Standard :code:`<execution>` header for more information.

Reviewer: ```suggestion
Please see "Execution policies" in the `C++ Standard`_ for more information.
```

  };

An object of the ``device_policy`` type is associated with a ``sycl::queue`` that is used
to run algorithms on a DPC++ compliant device. When an algorithm runs with ``device_policy``
it is capable of processing SYCL buffers (passed via :code:`oneapi::dpl::begin/end`),

> data in the host memory or data in Unified Shared Memory (USM), inlcuding USM device memory.

Author: ```suggestion
data in the host memory and data in Unified Shared Memory (USM), including USM device memory.
```

> data in the host memory or data in Unified Shared Memory (USM), inlcuding USM device memory.

Author: done

When data is placed in the host memory, it shall be passed to the algorithms as random access

> iterators. oneDPL algorithm might do implicit transfering data to a device before execution.

Reviewer: I would use something like the following:

> Data placed in the host memory and USM can only be passed to oneDPL algorithms as pointers and random access iterators. The way to transfer data from the host memory to a device and back is unspecified; per-element data movement to/from a temporary storage is a possible valid implementation.

The wording is discussable, but the two key ideas are:
- data in the host memory and in USM can be passed to algorithms in the same ways
- an implementation may, if needed, implicitly copy data from host memory to a temporary storage and back

Author: done

> When data is placed in the host memory, it shall be passed to the algorithms as random access
> iterators. oneDPL algorithm might do implicit transfering data to a device before execution.

Author: ```suggestion
Data placed in the host memory and USM can only be passed to oneDPL algorithms
as pointers and random access iterators. The way to transfer data from the host memory
to a device and back is unspecified; per-element data movement to/from a temporary storage
is a possible valid implementation.
```
